src/main/java/com/datapipe/jenkins/vault/VaultBindingStep.java

```diff
@@ -0,0 +1,332 @@
+import org.jenkinsci.plugins.workflow.steps.Step;
+import org.jenkinsci.plugins.workflow.steps.StepContext;
+import org.jenkinsci.plugins.workflow.steps.StepDescriptor;
+import org.jenkinsci.plugins.workflow.steps.StepExecution;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
+
+public class VaultBindingStep extends Step {
+
+    private VaultConfiguration configuration;
+    private List<VaultSecret> vaultSecrets;
+
+    @DataBoundConstructor
+    public VaultBindingStep(@CheckForNull List<VaultSecret> vaultSecrets) {
+        this.vaultSecrets = vaultSecrets;
+    }
+
+    public List<VaultSecret> getVaultSecrets() {
+        return vaultSecrets;
+    }
+
+    @DataBoundSetter
+    public void setConfiguration(VaultConfiguration configuration) {
+        this.configuration = configuration;
+    }
+
+    public VaultConfiguration getConfiguration() {
+        return configuration;
+    }
+
+    @Override
+    public StepExecution start(StepContext context) throws Exception {
+        return new Execution(this, context);
+    }
+
+    private static final class Execution extends GeneralNonBlockingStepExecution {
+
+        private transient VaultBindingStep step;
+
+        public Execution(VaultBindingStep step, StepContext context) {
+            super(context);
+            this.step = step;
+        }
+
+        @Override
+        public boolean start() throws Exception {
+            run(this::doStart);
+            return false;
+        }
+
+        private void doStart() throws Exception {
+            Run<?, ?> run = getContext().get(Run.class);
+            TaskListener listener = getContext().get(TaskListener.class);
+            EnvVars envVars = getContext().get(EnvVars.class);
+
+            Map<String, String> overrides = new HashMap<String, String>();
+
+            VaultConfiguration config = pullAndMergeConfiguration(run);
+            String url = config.getVaultUrl();
+
+            if (StringUtils.isBlank(url)) {
+                throw new VaultPluginException(
+                    "The vault url was not configured - please specify the vault url to use.");
+            }
+
+            VaultConfig vaultConfig = config.getVaultConfig();
+            VaultCredential credential = config.getVaultCredential();
+            if (credential == null) {
+                credential = retrieveVaultCredentials(run, config);
+            }
+
+            String prefixPath = StringUtils.isBlank(config.getPrefixPath())
+                ? ""
+                : Util.ensureEndsWith(envVars.expand(config.getPrefixPath()), "/");
+
+            VaultAccessor vaultAccessor = new VaultAccessor();
+            vaultAccessor.setConfig(vaultConfig);
+            vaultAccessor.setCredential(credential);
+            vaultAccessor.setMaxRetries(config.getMaxRetries());
+            vaultAccessor.setRetryIntervalMilliseconds(config.getRetryIntervalMilliseconds());
+            vaultAccessor.init();
```

Comment:
This feel like duplicated 3 times 🤔

Reply:
Sadly yes, I am not sure if I should start a refactoring to get rid of the duplication with a fix or this should be handled in a separate "cleanup" issue.

Reply:
Feel free to include the refactoring in this PR as the duplication is getting somewhat out of hand in this plugin already 😓

Reply:
Moved common code into VaultAccessor for now